iwyu_output.cc

```diff
@@ -578,14 +578,19 @@ void IwyuFileInfo::ReportFullSymbolUse(SourceLocation use_loc,
                                        const char* comment) {
   if (decl) {
     // Since we need the full symbol, we need the decl's definition-site.
-    // But only if we're not defining the function, in which case we want to use
-    // all its preceding declarations, and not try to canonicalize.
-    if (!(flags & UF_FunctionDfn)) {
-      decl = GetDefinitionAsWritten(decl);
+    // Also, by default we canonicalize the location, using GetLocation.
+    const NamedDecl* report_decl = GetDefinitionAsWritten(decl);
+    SourceLocation report_decl_loc = GetLocation(report_decl);
+
+    // However, if we're defining the function or we are targeting an explicit
+    // instantiation, we want to use it as-is and not try to canonicalize at all.
+    if ((flags & (UF_FunctionDfn | UF_ExplicitInstantiation)) != 0) {
```

Comment:
I wonder if it would be possible to turn this into an if/else block instead, something like:
```c++
SourceLocation decl_loc;

if (flags & (UF_FunctionDfn | UF_ExplicitInstantiation) == 0) {
    // Rationale here...
    decl = GetDefinitionAsWritten(decl);
    decl_loc = GetLocation(report_decl);
} else {
    // ... and here
    decl_loc = report_decl->getLocation();
}
```
That way it looks more balanced (even if it's annoying with the `decl_loc` declaration outside the block.